torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,295 @@
+        tensordict = tensordict.clone(recurse=False)
+
+        # prepare tensordict: remove time in batch dimensions
+        tensordict.batch_size = tensordict.batch_size[:1]
+        # take the first tensor for prev_posterior_state and prev_belief
+        tensordict["prev_posterior_state"] = torch.zeros_like(
+            tensordict["prev_posterior_state"][:, 0]
+        )
+        tensordict["prev_belief"] = torch.zeros_like(tensordict["prev_belief"][:, 0])
+        tensordict["true_reward"] = tensordict["reward"]
+        del tensordict["reward"]
+
+        tensordict = self.world_model(tensordict)
+        # compute model loss
+        kl_loss = self.kl_loss(
+            tensordict.get("prior_means"),
+            tensordict.get("prior_stds"),
+            tensordict.get("posterior_means"),
+            tensordict.get("posterior_stds"),
+        )
+        reco_loss = (
+            distance_loss(
+                tensordict.get("pixels"),
+                tensordict.get("reco_pixels"),
+                self.reco_loss,
+            )
+            .sum((-1, -2, -3))
+            .mean()
+        )
+        reward_loss = distance_loss(
+            tensordict.get("true_reward"),
+            tensordict.get("reward"),
+            self.reward_loss,
+        ).mean()
+        loss = (
+            self.lambda_kl * kl_loss
+            + self.lambda_reco * reco_loss
+            + self.lambda_reward * reward_loss
+        )
+        return (
+            TensorDict(
+                {
+                    "loss_world_model": loss,
+                    "kl_model_loss": kl_loss,
+                    "reco_model_loss": reco_loss,
+                    "reward_model_loss": reward_loss,
+                },
+                [],
+            ),
+            tensordict.detach(),
+        )
+
+    def kl_loss(
+        self,
+        prior_mean: torch.Tensor,
+        prior_std: torch.Tensor,
+        posterior_mean: torch.Tensor,
+        posterior_std: torch.Tensor,
+    ) -> torch.Tensor:
+        kl = (
+            torch.log(prior_std / posterior_std)
+            + (posterior_std ** 2 + (prior_mean - posterior_mean) ** 2)
+            / (2 * prior_std ** 2)
+            - 0.5
+        )
+        if self.inversed_free_nats:
+            kl = kl.sum(-1).mean().clamp_min(self.free_nats)
+        else:
+            kl = kl.sum(-1).clamp_min(self.free_nats).mean()
+        return kl
+
+
+class DreamerActorLoss(LossModule):
+    """Dreamer Actor Loss
+
+    Computes the loss of the dreamer actor. The actor loss is computed as the negative lambda return average.
+
+    Reference: https://arxiv.org/abs/1912.01603
+
+    Args:
+        actor_model (TensorDictModule): the actor model.
+        value_model (TensorDictModule): the value model.
+        model_based_env (DreamerEnv): the model based environment.
+        cfg (DictConfig): the config file.
+        gamma (float, optional): the discount factor.
```

Comment:
let's not forget about default values, it's important when users call `help(class)`

Reply:
lambda is also a discount factor, no?